Fix backout

lutien opened the pull request.

jgraham commented:
So, I think this is correct, but I'm not sure it's sufficient to fix the entire problem.

https://github.com/mozilla/wpt-sync/blob/master/sync/upstream.py#L722 is the main entry point for updating for pushes to gecko repos e.g. autoland.

We call `remove_complete_backouts` first, but all that does is removes commits that are pushed and backed out in the same set of commits, as processed by the sync bot. But that's actually a rather uncommon case, the more normal thing is that the commit and backout aren't processed at the same time. In that case `remove_complete_backouts` doesn't do the right thing (because we don't want to process every gecko commit going back indefinitely). Instead if there's a backout in the commit range we're processing, we end up in `updates_for_backout` and I think [the bug](https://github.com/mozilla/wpt-sync/blob/master/sync/upstream.py#L688-L691) in _that_ is the assumption that if it finds a sync that was backed out by a specific backout commit then it must be done with that commit. It doesn't allow for the possibility that a single backout commit can undo multiple syncs.

lutien commented:
The thing is that `remove_complete_backouts` runs also in [filter_commits](https://github.com/mozilla/wpt-sync/blob/master/sync/upstream.py#L68), which runs when we access [commits](https://github.com/mozilla/wpt-sync/blob/master/sync/sync.py#L179). As far as I understood, how it works in the test, which I added: `remove_complete_backouts` runs with only backout commit, then it goes into `updates_for_backout` creates correct updates for both syncs, where it adds backed out commit to both of them, then we go [here](https://github.com/mozilla/wpt-sync/blob/master/sync/upstream.py#L899), when we access commits of sync we have an actual commit and backout and if we filter correctly, when we get [here](https://github.com/mozilla/wpt-sync/blob/master/sync/upstream.py#L853), PR gets invalidated. 
Now coming back to the issue with `updates_for_backout`, I guess it all works fine in my test because backout is applied to both syncs in one go, so we don't get the case when the backout commit is already applied to the sync. But true, maybe it can happen, and then we don't need to apply it to this sync again, but we don't want to just exit without checking other syncs, so I've added the commit to continue the loop in this case. Let me know what you think.

lutien commented on sync/upstream.py:
```diff
@@ -687,9 +687,8 @@ def updates_for_backout(git_gecko: Repo,
-                backed_out_commit_shas = set()
-                return {}, {}
+                # This commit was already processed for this sync
+                continue
```
yeah, you're right about the removing from `backed_out_commit_shas`, I've updated this.
Regarding the scenario when S1 is merged, I guess I'm missing something, but shouldn't S1 then have status `complete` and because of that excluded from all the backout updates, so it shouldn't come to the "create a sync" point?

jgraham replied:
It's exactly because it's merged that we get to the create a sync point. The whole "create a new sync for the backout commit" is designed to handle the case where e.g. something is backed out of central _after_ the corresponding upstream sync has already been merged. In that case we can't do anything with the original PR (because GitHub quite rightly doesn't allow adding new commits to a merged PR) so instead we want to create an entirely new sync/PR for the backout.

lutien replied:
Ok, I think I'm up to speed with what you're saying :)
Just to reiterate, basically the problem is that in the case you described, we're creating a new sync with only the backout commit, which works fine for one bug backout, but doesn't for multiple bugs, where we should revert only the commits, which belong to the bug, which was merged already.
But so far I don't have a better idea, than, as you said, creating a sync with reverted commits for the bug. So, I guess, I'll try to sketch something.
